Feature #172453733 – Add external resource links

lingyun1010 opened the pull request:
The logic for adding external resources discussed with curators is that we add all links from one resource and if there are multiple resources available we follow the priority GEO > ENA and EGA > ENA. We only look for accessions from `comment[secondaryAccession]` in IDF files. The accessions for GEO resources starts from `GSE`, for EGA resources starts from `EGA` or `EGAS`, while the accessions for ENA resources are a bit complicated, refer to the [ENA link](https://ena-docs.readthedocs.io/en/latest/submit/general-guide/accessions.html).

Adding resource links and icons are in the [bulk feature branch](https://github.com/ebi-gene-expression-group/atlas-web-bulk/pull/64).

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java:
```diff
@@ -66,7 +70,7 @@ public Experiment(@NotNull ExperimentType type,
                       @NotNull String accessKey) {
         checkArgument(isNotBlank(accession));
-        checkArgument(isNotBlank(description));
+        checkArgument(!description.isEmpty());
```
`String::isEmpty()` is not null-safe and will throw a NPE if `description` is null. The reasoning is similar for `accession`. Remember that the `@NotNull` annotation is just a hint for your IDE to signal if there are potential problems, but there’s no enforcing at run-time that `description` won’t be null.

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java:
```diff
@@ -25,6 +26,7 @@
 // There's also a title which is fetched from the IDF file.
 public abstract class Experiment<R extends ReportsGeneExpression> implements Serializable {
     private final ExperimentType type;
+    private final List<String> secondaryAccession;
```
Use `ImmutableList`, please.

Also, since this is a collection, `secondaryAccessions` is a better name. Also, unless you’re accessing this field by index (i.e. `secondaryAccessions.get(...)`) an `ImmutableSet` is a better choice.

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java:
```diff
@@ -75,6 +79,7 @@ public Experiment(@NotNull ExperimentType type,
 
         this.type = type;
         this.accession = accession;
+        this.secondaryAccession = secondaryAccession;
```
When assigning mutable variables to fields of an object, you should always create a copy. Otherwise you risk that changes in the object outside will be reflected inside the instance that holds a reference. See the other non-primitive fields below.

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java:
```diff
@@ -96,6 +101,9 @@ public Experiment(@NotNull ExperimentType type,
     }
 
+    @NotNull
+    public List<String>  getSecondaryAccession() { return secondaryAccession; }
```
When you change the field to an immutable collection remember to narrow down the returned type here.

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/testutils/MockExperiment.java:
```diff
@@ -150,7 +150,7 @@ public static BaselineExperiment createBaselineExperiment(String accession,
                 RNASEQ_MRNA_BASELINE,
                 accession,
-                "",
+                ImmutableList.of(""),
```
Express the absence of secondary accessions with an empty collection rather than by an singleton with an empty string. Modify the other mock experiments accordingly.

alfonsomunozpomer commented on src/test/java/uk/ac/ebi/atlas/experimentpage/LinkToGeoEnaEgaTest.java:
```diff
@@ -0,0 +1,236 @@
+                .isEqualTo(new LinkToGeoEnaEga.ProteomicsBaseline().contentType())
+                .isEqualTo(new LinkToGeoEnaEga.Differential().contentType())
+                .isEqualTo(new LinkToGeoEnaEga.Microarray().contentType())
+                .isEqualTo(SUPPLEMENTARY_INFORMATION);
```
This is a nice one. :medal_sports:

alfonsomunozpomer commented on src/test/java/uk/ac/ebi/atlas/experimentpage/LinkToGeoEnaEgaTest.java:
```diff
@@ -0,0 +1,236 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static uk.ac.ebi.atlas.model.download.ExternallyAvailableContent.ContentType.SUPPLEMENTARY_INFORMATION;
+
+class LinkToGeoEnaEgaTest {
```
Your `LinkToGeoEnaEga` class can produce links to different domains, so you should also test with `allMatch` that the links point at their respective hosts depending on the secondary accession. This isn’t the case in `LinkToArrayExpress`, therefore only the last segment of the URL is tested.

Read my general comments in the review, though.

alfonsomunozpomer commented on src/test/java/uk/ac/ebi/atlas/model/experiment/ExperimentBuilder.java:
```diff
@@ -53,7 +53,8 @@
     String experimentAccession = generateRandomExperimentAccession();
-    String secondaryExperimentAccession = RNG.nextBoolean() ? generateRandomPrideExperimentAccession() : "";
+    List<String> secondaryExperimentAccessions = RNG.nextBoolean() ? ImmutableList.of(generateRandomPrideExperimentAccession())
+            : ImmutableList.of("");
```
Remember to have an empty collection if there are no secondary accessions.

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/experimentimport/idf/IdfParser.java:
```diff
@@ -8,6 +9,7 @@
 import uk.ac.ebi.atlas.resource.DataFileHub;
 
 import java.util.Arrays;
+import java.util.HashSet;
```
I think you can remove this import.

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/experimentimport/idf/IdfParserOutput.java:
```diff
@@ -19,13 +21,13 @@
 
     public IdfParserOutput(String title,
-                           String secondaryAccession,
+                           ImmutableSet<String> secondaryAccessions,
```
Remember: your methods should **accept** a type as **broad** as possible, see the examples below:

```java
public ClassA(Collection<T> things) {
    this.things = ImmutableSet.copyOf(things);
}
```

```java
public ClassA(ImmutableCollection<T> things) {
    this.things = things;
}
```

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/experimentimport/idf/IdfParserOutput.java:
```diff
@@ -36,8 +38,8 @@ public String getTitle() {
 
-    public String getSecondaryAccession() {
-        return secondaryAccession;
+    public Collection<String> getSecondaryAccessions() {
```
Secondly, your methods shoud **return** type as **specific** as possible (`ImmutableSet<String> here`).

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/experimentpage/link/GenerateResourceLinks.java:
```diff
@@ -0,0 +1,50 @@
+                                                                   String regex,
+                                                                   UriBuilder uriBuilder,
+                                                                   Function<String, ExternallyAvailableContent.Description> createIcon) {
+        var secondaryccessions = experiment.getSecondaryAccessions().stream()
+                .filter(accession -> accession.matches(regex))
+                .collect(toImmutableList());
+
+        return secondaryccessions.stream()
+                .parallel()
+                .map(accession -> Pair.of(uriBuilder.build(accession), accession))
+                .filter(uriAccession -> isUriValid(uriAccession.getLeft()))
+                .map(uriAccession -> new ExternallyAvailableContent(
+                        uriAccession.getLeft().toString(),
+                        createIcon.apply(uriAccession.getRight())))
+                .collect(toImmutableList());
```
You can merge these two streams and return directly.

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/experimentpage/link/GenerateResourceLinks.java:
```diff
@@ -0,0 +1,50 @@
+public class GenerateResourceLinks {
+    private static final WebClient webClient = WebClient.create();
+
+    public static Collection<ExternallyAvailableContent> getLinks(Experiment<?> experiment,
```
Return a specific type.

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/experimentpage/link/GenerateResourceLinks.java:
```diff
@@ -0,0 +1,50 @@
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class GenerateResourceLinks {
```
I like this class!

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/experimentpage/link/GenerateResourceLinks.java:
```diff
@@ -0,0 +1,50 @@
+    public static Collection<ExternallyAvailableContent> getLinks(Experiment<?> experiment,
+                                                                   String regex,
+                                                                   UriBuilder uriBuilder,
+                                                                   Function<String, ExternallyAvailableContent.Description> createIcon) {
```
:100:

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/experimentpage/link/LinkToEga.java:
```diff
@@ -0,0 +1,93 @@
+        }
+    }
+
+    private static boolean isUriValid(@NotNull URI uri) {
+        try {
+            return !webClient
+                    .get()
+                    .uri(uri)
+                    .exchange()
+                    .block()
+                    .statusCode()
+                    .isError();
+        } catch (Exception e) {
+            return false;
+        }
+    }
```
No need for this now, here and elsewhere. Just in `GenerateResourceLinks`.

alfonsomunozpomer commented on src/main/java/uk/ac/ebi/atlas/model/experiment/Experiment.java:
```diff
@@ -25,6 +25,7 @@
 // There's also a title which is fetched from the IDF file.
 public abstract class Experiment<R extends ReportsGeneExpression> implements Serializable {
     private final ExperimentType type;
+    private final Collection<String> secondaryAccessions;
```
The field type should also be specific.

lingyun1010 commented:
> There’s one unaddressed comment (the unnecessary `HashSet` import). There are still some remaining `WebClient` fields and `isUriValid` methods in your classes that should also be removed.
> 
> As the last step [to dry your code](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), notice that all the static subclasses of your `LinkTo...` classes have exactly the same implementation. E.g.:
> 
> ```java
> ...
> @Component
> public static class ProteomicsBaseline extends LinkToEga<BaselineExperiment> {
>     @Override
>     public Collection<ExternallyAvailableContent> get(BaselineExperiment experiment) {
>         return GenerateResourceLinks.getLinks(experiment, "EGA.*", EGA_URI_BUILDER, createIconForEga);
>     }
> }
> 
> @Component
> public static class RnaSeqBaseline extends LinkToEga<BaselineExperiment> {
>     @Override
>     public Collection<ExternallyAvailableContent> get(BaselineExperiment experiment) {
>         return GenerateResourceLinks.getLinks(experiment, "EGA.*", EGA_URI_BUILDER, createIconForEga);
>     }
> }
> ...
> ```
> 
> You can refactor all of them to a common `get` method. Have a look at `LinkToArrayExpress` for an example. Also, maybe you can modify `LinkToArrayExpress` to use methods in `GenerateResourceLinks`?

`LinkToArrayExpress` use `getAccession()` so I do not think i

lingyun1010 commented:
Thanks for the comments. As `LinkToArrayExpress` is using different accession items than the others, I did not reformat the codes to use `GenerateResourceLinks`.
